examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+somewhat overloaded term, here we are talking about it in a Riemannian geometry
+sense.
+
+' In the `TangentSpace` interface, `a` is the type which can represent points
+on the manifold. `b` is the type which can represent points in a tangent space
+to the manifold.  Ideally `b` would be an associated type of `a` so that the
+compiler is able to infer the tangent space type from the manifold type, but
+this feature is not yet available in Dex.
+
+interface [VSpace b] TangentSpace a b
+  -- Given a point p on the manifold M and a point v in its tangent space TpM,
+  -- this returns another point q on the manifold.
+  -- Must satisfy tangentExponentialMap p zero == p
+  -- (tangentExponentialMap p) will be linearized around zero, so it is
+  -- important that any operations used have well-defined derivatives for inputs
+  -- p, zero.
+  tangentExponentialMap: a -> b -> a
+  -- (tangentLogarithmMap p) is the pointwise inverse to
+  -- (tangentExponentialMap p).
+  -- Must satisfy tangentLogarithmMap p p == zero
+  -- (tangentLogarithmMap p) will be linearized around p, so it is important
+  -- that any operations used have well-defined derivatives for inputs p, p.
+  tangentLogarithmMap: a -> a -> b
+
+
+'### Differentiation functions
+
+'We can then define a version of the Dex `linearize` function which works on
+manifolds.  To see intuitively why this works, consider the following diagram.
+We would like to obtain a linearized version of the function "across the top" of
+the diagram, and we can form this by linearizing the function formed from the
+bottom three sides of the square.
+
+:html
+   "<img
+     src=\"https://user-images.githubusercontent.com/33986545/122416078-5a23ef00-cf80-11eb-83d5-d0d5986e8d20.png\"
+     width=\"100%\">"
+
+
+def manifoldLinearize [TangentSpace a a', TangentSpace b b', Add a']
+    (f: a -> b) (x: a) : (b & a' --o b') = 
+  y = f x
+  tangentSpaceMapping : a' -> b' =
+    \ v . tangentLogarithmMap y $ f $ tangentExponentialMap x v
+  rawLinearize = linearize tangentSpaceMapping zero
+  (y, snd rawLinearize)
+
+' A full proof that this is in fact a correct formulation is a bit more
+involved, so we leave it out here. The general idea is to use the differentiable
+curves formulation of the tangent space, and to treat the tangent space at a 
+point $x$ as both the tangent space, and as a local chart.
+
+' Before we can define `grad` for manifolds, we need to create the trivial
+instance of a TangentSpace for vector spaces. In this case, we view the tangent
+space at $x$ to be just a copy of the vector space, centred on $x$.
+
+instance [VSpace a] TangentSpace a a
+  tangentExponentialMap = \ p v . p + v
+  tangentLogarithmMap = \ p q . q - p
+
+' Now we can define a version of the Dex `grad` function which works on
+manifolds.
+
+def manifoldGrad [TangentSpace a a', VSpace a'] (f: a -> Float) (x: a) : a' = 
+  linearized: (Float & (a' --o Float)) = manifoldLinearize f x
+  transposeLinear (snd linearized) 1.0
```

Comment:
I love how the decomposition of reverse-mode AD lets you modify the linearization part here while reusing standard transposition!

Reply:
😁